sympy/ntheory/qs.py

```diff
@@ -0,0 +1,391 @@
+from sympy.core.numbers import igcd, mod_inverse
+from sympy.ntheory.residue_ntheory import _sqrt_mod_prime_power
+from sympy.ntheory import isprime
+from math import log, sqrt
+import random
+
+class sieve_Polynomial:
+    def __init__(self, modified_coeff=[], a=None, b=None):
+        """This class denotes the seive polynomial.
+        If `g(x) = (a*x + b)**2 - N. g(x)` can be expanded
+        to `a*x**2 + 2*a*b*x + b**2 - N`, so the coefficient
+        is stored in the form `[a**2, 2*a*b, b**2 - N]`. This
+        ensures faster `eval` method because we dont have to
+        perform `a**2, 2*a*b, b**2` every time we call the
+        `eval` method. As multiplication is more expensive
+        than addition, by using modified_coefficient we get
+        a faster seiving process.
+
+        Parameters:
+        ===========
+
+        modified_coeff : modified_coefficient of sieve polynomial
+        a : parameter of the sieve polynomial
+        b : parameter of the sieve polynomial
+        """
+        self.modified_coeff = modified_coeff
+        self.a = a
+        self.b = b
+
+    def eval(self, x):
+        """
+        Compute the value of the sieve polynomial at point x.
+
+        Parameters:
+        ===========
+
+        x : Integer parameter for sieve polynomial
+        """
+        ans = 0
+        for coeff in self.modified_coeff:
+            ans *= x
+            ans += coeff
+        return ans
+
+class factor_base_elem:
+    """This class stores an element of te factor_base.
+    """
+    def __init__(self, prime, tmem_p, log_p):
+        """
+        Initialization of factor_base_elem.
+
+        Parameters:
+        ===========
+
+        prime : prime number of the factor_base
+        tmem_p : Integer square root of x**2 = n mod prime
+        log_p : Compute Natural Logarithm of the prime
+        """
+        self.prime = prime
+        self.tmem_p = tmem_p
+        self.log_p = log_p
+        self.soln1 = None
+        self.soln2 = None
+        self.a_inv = None
+        self.B_ainv = None
+
+
+def gen_factor_base(B, n):
+    """Generate factor_base for Quadratic Sieve. The factor_base
+    consists of all the the points whose legendre_symbol(n, p) == 1
+    and p < B.
+
+    Parameters:
+    ===========
+
+    B : upper bound of the factor_base
+    n : integer to be factored
+    """
+    from sympy import prime
+    factor_base = []
+    i = 1
+    while len(factor_base) < B:
+        p = prime(i)
+        i += 1
```

Comment:
What does i refer to here?

Reply:
it is just an increment. instead of taking primerange i used number of primes required.